<a href="https://colab.research.google.com/github/kcw0331/Environmental-Health-Big-Data-Analysis/blob/main/6%EC%A3%BC%EC%B0%A8/4-inspecting_dataframes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Inspecting a DataFrame Object

## About the Data
In this notebook, we will be working with earthquake data from September 18, 2018 - October 13, 2018 (obtained from the US Geological Survey (USGS) using the [USGS API](https://earthquake.usgs.gov/fdsnws/event/1/))

## Setup
We will be working with the `data/earthquakes.csv` file again, so we need to handle our imports and read it in.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
cd /content/drive/MyDrive/Hands-On-Data-Analysis-with-Pandas-2nd-edition/ch_02

/content/drive/MyDrive/Hands-On-Data-Analysis-with-Pandas-2nd-edition/ch_02


In [3]:
import numpy as np
import pandas as pd

df = pd.read_csv('data/earthquakes.csv')  #데이터를 데이터 프레임으로 불러와준다.

## Examining dataframes
### Is it empty?

In [4]:
df.empty

False

* 비어있는게 있는지 확인을 해주는 것이다. False로 나오기 때문에 비어있는 것을 없다는 것이다.

### What are the dimensions?

In [5]:
df.shape

(9332, 26)

* 데이터 프레임의 차원이 어떻게 되냐는 말이다. 그래서 row는 9332가 있고 feature의 개수는 26개가 있다는 것이다.

### What columns do we have?
We know there are 26 columns, but what are they? Let's use the `columns` attribute to see:

In [6]:
df.columns

Index(['alert', 'cdi', 'code', 'detail', 'dmin', 'felt', 'gap', 'ids', 'mag',
       'magType', 'mmi', 'net', 'nst', 'place', 'rms', 'sig', 'sources',
       'status', 'time', 'title', 'tsunami', 'type', 'types', 'tz', 'updated',
       'url'],
      dtype='object')

* columns들이 어떠한 것들이 있는지 df.columns로 확인을 해준다.

### What does the data look like?
View rows from the top with `head()`:

In [7]:
df.head(2)

alert  cdi  ...        updated                                                url
0   NaN  NaN  ...  1539475395144  https://earthquake.usgs.gov/earthquakes/eventp...
1   NaN  NaN  ...  1539475253925  https://earthquake.usgs.gov/earthquakes/eventp...

[2 rows x 26 columns]

View rows from the bottom with `tail()`. Let's view 2 rows:

In [8]:
df.tail(2)

alert  ...                                                url
9330   NaN  ...  https://earthquake.usgs.gov/earthquakes/eventp...
9331   NaN  ...  https://earthquake.usgs.gov/earthquakes/eventp...

[2 rows x 26 columns]

* tail을 사용해서 밑에 있는 2개를 확인해 줄 수 있다.

*Tip: we can modify the display options in order to see more columns:*

```python
# check the max columns setting
>>> pd.get_option('display.max_columns')
20

# set the max columns to show when printing the dataframe to 26
>>> pd.set_option('display.max_columns', 26)
# OR
>>> pd.options.display.max_columns = 26

# reset the option
>>> pd.reset_option('display.max_columns')

# get information on all display settings
>>> pd.describe_option('display')
```

*More information can be found in the documentation [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html).*

### What data types do we have?

In [9]:
df.dtypes

alert       object
cdi        float64
code        object
detail      object
dmin       float64
felt       float64
gap        float64
ids         object
mag        float64
magType     object
mmi        float64
net         object
nst        float64
place       object
rms        float64
sig          int64
sources     object
status      object
time         int64
title       object
tsunami      int64
type        object
types       object
tz         float64
updated      int64
url         object
dtype: object

### Getting extra info and finding nulls

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9332 entries, 0 to 9331
Data columns (total 26 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   alert    59 non-null     object 
 1   cdi      329 non-null    float64
 2   code     9332 non-null   object 
 3   detail   9332 non-null   object 
 4   dmin     6139 non-null   float64
 5   felt     329 non-null    float64
 6   gap      6164 non-null   float64
 7   ids      9332 non-null   object 
 8   mag      9331 non-null   float64
 9   magType  9331 non-null   object 
 10  mmi      93 non-null     float64
 11  net      9332 non-null   object 
 12  nst      5364 non-null   float64
 13  place    9332 non-null   object 
 14  rms      9332 non-null   float64
 15  sig      9332 non-null   int64  
 16  sources  9332 non-null   object 
 17  status   9332 non-null   object 
 18  time     9332 non-null   int64  
 19  title    9332 non-null   object 
 20  tsunami  9332 non-null   int64  
 21  type     9332 

## Describing and Summarizing
### Get summary statistics

In [11]:
df.describe()

cdi         dmin  ...           tz       updated
count  329.000000  6139.000000  ...  9331.000000  9.332000e+03
mean     2.754711     0.544925  ...  -451.990140  1.538537e+12
std      1.010637     2.214305  ...   231.752571  6.564135e+08
min      0.000000     0.000648  ...  -720.000000  1.537230e+12
25%      2.000000     0.020425  ...  -540.000000  1.537996e+12
50%      2.700000     0.059050  ...  -480.000000  1.538621e+12
75%      3.300000     0.177250  ...  -480.000000  1.539110e+12
max      8.400000    53.737000  ...   720.000000  1.539537e+12

[8 rows x 13 columns]

* summary statistic을 얻고 싶을 때 사용을 해준다.

Specifying the 5<sup>th</sup> and 95<sup>th</sup> percentile:

In [12]:
df.describe(percentiles=[0.05, 0.95])

cdi         dmin  ...           tz       updated
count  329.000000  6139.000000  ...  9331.000000  9.332000e+03
mean     2.754711     0.544925  ...  -451.990140  1.538537e+12
std      1.010637     2.214305  ...   231.752571  6.564135e+08
min      0.000000     0.000648  ...  -720.000000  1.537230e+12
5%       2.000000     0.005491  ...  -600.000000  1.537387e+12
50%      2.700000     0.059050  ...  -480.000000  1.538621e+12
95%      4.300000     2.678900  ...   -60.000000  1.539400e+12
max      8.400000    53.737000  ...   720.000000  1.539537e+12

[8 rows x 13 columns]

Describe specific data types:

In [13]:
df.describe(include=np.object)

alert  ...                                                url
count      59  ...                                               9332
unique      2  ...                                               9332
top     green  ...  https://earthquake.usgs.gov/earthquakes/eventp...
freq       58  ...                                                  1

[4 rows x 13 columns]

* (include=np.object)이렇게 해주게 되면 categorical 데이터에 대해서 살펴볼 수 있다.

Or describe all of them:

In [14]:
df.describe(include='all')

alert  ...                                                url
count      59  ...                                               9332
unique      2  ...                                               9332
top     green  ...  https://earthquake.usgs.gov/earthquakes/eventp...
freq       58  ...                                                  1
mean      NaN  ...                                                NaN
std       NaN  ...                                                NaN
min       NaN  ...                                                NaN
25%       NaN  ...                                                NaN
50%       NaN  ...                                                NaN
75%       NaN  ...                                                NaN
max       NaN  ...                                                NaN

[11 rows x 26 columns]

This works on columns also:

In [15]:
df.felt.describe()

count    329.000000
mean      12.310030
std       48.954944
min        0.000000
25%        1.000000
50%        2.000000
75%        5.000000
max      580.000000
Name: felt, dtype: float64

* felt variable에 대한 정보를 얻고 싶을 때 describe를 사용해 준다.

There are methods for specific statistics as well. Here is a sampling of them:

| Method | Description | Data types |
| --- | --- | --- |
| `count()` | The number of non-null observations | Any |
| `nunique()` | The number of unique values | Any |
| `sum()` | The total of the values | Numerical or Boolean |
| `mean()` | The average of the values | Numerical or Boolean |
| `median()` | The median of the values | Numerical |
| `min()` | The minimum of the values | Numerical |
| `idxmin()` | The index where the minimum values occurs | Numerical |
| `max()` | The maximum of the values | Numerical |
| `idxmax()` | The index where the maximum value occurs | Numerical |
| `abs()` | The absolute values of the data | Numerical |
| `std()` | The standard deviation | Numerical |
| `var()` | The variance |  Numerical |
| `cov()` | The covariance between two `Series`, or a covariance matrix for all column combinations in a `DataFrame` | Numerical |
| `corr()` | The correlation between two `Series`, or a correlation matrix for all column combinations in a `DataFrame` | Numerical |
| `quantile()` | Calculates a specific quantile | Numerical |
| `cumsum()` | The cumulative sum | Numerical or Boolean |
| `cummin()` | The cumulative minimum | Numerical |
| `cummax()` | The cumulative maximum | Numerical |

For example, finding the unique values in the `alert` column:

* 관심있는 학생들은 위에 있는 거를 한번 해보라고 말씀하심.

In [16]:
df.alert.unique()

array([nan, 'green', 'red'], dtype=object)

We can then use `value_counts()` to see how many of each unique value we have:

In [17]:
df.alert.value_counts()

green    58
red       1
Name: alert, dtype: int64

Note that `Index` objects also have several methods to help describe and summarize our data:

| Method | Description |
| --- | --- |
| `argmax()`/`argmin()` | Find the location of the maximum/minimum value in the index |
| `equals()` | Compare the index to another `Index` object for equality |
| `isin()` | Check if the index values are in a list of values and return an array of Booleans |
| `max()`/`min()` | Find the maximum/minimum value in the index |
| `nunique()` | Get the number of unique values in the index |
| `to_series()` | Create a `Series` object from the index |
| `unique()` | Find the unique values of the index |
| `value_counts()`| Create a frequency table for the unique values in the index |

* index object에 대한게 나와있는데 다음에 자세히 본다고 하심.

<hr>
<div>
    <a href="./3-making_dataframes_from_api_requests.ipynb">
        <button style="float: left;">&#8592; Previous Notebook</button>
    </a>
    <a href="./5-subsetting_data.ipynb">
        <button style="float: right;">Next Notebook &#8594;</button>
    </a>
</div>
<br>
<hr>